Sorry for the bad initial PR :( Retake:

Issue #26043 is about that rustc invalidly will consider its own libc as a candidate for `--extern libc=test.rlib`.

That was accidentally fixed in `nightly-2020-07-31`. With `nightly-2020-07-30` (the nightly before the fix) we see this:

```
$ RUSTC_LOG=debug rustc +bisector-nightly-2020-07-30-x86_64-unknown-linux-gnu test.rs --extern libc=test.rlib
[...]
[INFO  rustc_metadata::creader] resolving crate `libc`
[...]
[INFO  rustc_metadata::creader] resolved crates:
[...]
[INFO  rustc_metadata::creader]   name: libc
[INFO  rustc_metadata::creader]   cnum: 6
[INFO  rustc_metadata::creader]   hash: 5cb655619dfdc51f
[INFO  rustc_metadata::creader]   reqd: Explicit
[INFO  rustc_metadata::creader]    rlib: /home/martin/.rustup/toolchains/bisector-nightly-2020-07-30-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-34294dcda38da0eb.rlib
[...]
error[E0658]: use of unstable library feature 'rustc_private': this crate is being loaded from the sysroot, an unstable location; did you mean to load this crate from crates.io via `Cargo.toml` instead?
 --> test.rs:1:1
  |
1 | extern crate libc;
  | ^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #27812 <https://github.com/rust-lang/rust/issues/27812> for more information
  = help: add `#![feature(rustc_private)]` to the crate attributes to enable

error: aborting due to previous error

For more information about this error, try `rustc --explain E0658`.
```

It is clear that `libc` is resolved wrongly. Both based on the debug log and based on the emitted user-facing error message. With `nightly-2020-07-31` we instead see this:

```
$ RUSTC_LOG=debug rustc +bisector-nightly-2020-07-31-x86_64-unknown-linux-gnu test.rs --extern libc=test.rlib
[...]
[INFO  rustc_metadata::creader] resolving crate `libc`
[INFO  rustc_metadata::creader] falling back to a load
error: extern location for libc does not exist: test.rlib
 --> test.rs:1:1
  |
1 | extern crate libc;
  | ^^^^^^^^^^^^^^^^^^

error: aborting due to previous error
```

We can see both that the `error` changed, and that the resolver failed to resolve `libc` at all. Given that we know in what nightly this changed, it most likely is https://github.com/rust-lang/rust/pull/74915 that accidentally fixed #26043.

For maintainability reasons, I think we should avoid writing a test that interprets `RUSTC_LOG` output. Instead, let's make use of that the error message changed. We can write a ui test that asserts on the "fixed" error.

I have pushed such a test now. Thank you for taking a look before. Would you mind taking another look please? Thanks!

r? Nilstrieb

@rustbot ready

--- a/tests/ui/imports/resolve-other-libc.rs
+++ b/tests/ui/imports/resolve-other-libc.rs
@@ -1,0 +1,14 @@
+// Regression test for https://github.com/rust-lang/rust/issues/26043
+
+// compile-flags: --extern libc=test.rlib
+
+// The error shall NOT be something similar to the following, because it
+// indicates that `libc` was wrongly resolved to `libc` shipped with the
+// compiler:
+//
+//   error[E0658]: use of unstable library feature 'rustc_private': \
+//           this crate is being loaded from the sysroot
+//
+extern crate libc; //~ ERROR: extern location for libc does not exist: test.rlib
+
+fn main() {}

--- a/tests/ui/imports/resolve-other-libc.stderr
+++ b/tests/ui/imports/resolve-other-libc.stderr
@@ -1,0 +1,8 @@
+error: extern location for libc does not exist: test.rlib
+  --> $DIR/resolve-other-libc.rs:12:1
+   |
+LL | extern crate libc;
+   | ^^^^^^^^^^^^^^^^^^
+
+error: aborting due to previous error
+

